### ISSUE TITLE
Flicker after synchronous state update with react-query

### ISSUE THREAD

**Issue by Liboul** [hello-pangea/dnd#424]
Hey 👋 Love the library, thanks for taking over `react-beautiful-dnd`, adding typing, supporting React 18 ... it's great!
I had this issue pop up while making the switch from `react-beautiful-dnd`, which makes me think it's a small regression

### Expected behavior

Making a synchronous state update through [react-query's setQueryData](https://tanstack.com/query/v4/docs/reference/QueryClient#queryclientsetquerydata) should not cause a flicker. By "flicker", I mean that the dropped element goes back to its original place then to the right place.

### Actual behavior

The dropped element goes back to its original place for a time

### Steps to reproduce

0. Implement `onDragEnd` to make an optimistic synchronous update, using react-query's setQueryData.
1. Drop an item.
2. Witness the flicker.

### Suggested solution?

Under the hood, react-query uses react's hook [useSyncExternalStore](https://reactjs.org/docs/hooks-reference.html#usesyncexternalstore). My best guess is that @hello-pangea/dnd does not wait for the state update before moving the dropped item back to its original place, as is described [here](https://github.com/hello-pangea/dnd/blob/main/docs/guides/responders.md)

> We try hard to ensure that an entire lifecycle is completed before a new one starts. If you find that not to be the case - it is a bug: please raise it!

### What version of `React` are you using?

18.2.0

### What version of `@hello-pangea/dnd` are you running?

16.0.0

### What browser are you using?

Chrome Version 105.0.5195.102

### Demo

Minimal use case implemented in this sandbox: https://codesandbox.io/s/react-beautiful-dnd-flicker-after-synchronous-update-with-reactquery-n207n1

https://user-images.githubusercontent.com/16024169/190988569-a145890f-1852-4c2b-9d65-d68fbb123b10.mov

Note: To enhance the issue on this dummy example, I faked a costly rendering time.

**Comment by Gorbus** [comment]
We're having exactly the same issue in the same configuration, any progress on this ?

**Comment by Liboul** [comment]
So far the only hack I have is to duplicate the state between react-query's cache and a local component state, but that's really not ideal, as it's easy to get out of sync and it requires more code.

I did try to `flushSync` in various places, to no avail.

**Comment by Xhale1** [comment]
Interesting. I ran into a similar issue when switching from react 17 to 18, which makes me think it's primarily an issue with how the library supports react 18.

The solution ended up being a simple flushSync call, I'll play with the code sandbox and see if I can find a solution in the morning. Thanks for the report!

**Comment by Gorbus** [comment]
Thanks a lot for having a look @Xhale1 , for us it's a very problematic regression. Ready to test if you have anything you think may help.

**Comment by Gorbus** [comment]
Any success on the subject by any chance @Xhale1 ? Still preventing us to upgrade to latest react unfortunately, but couldn't find a fix on my own :(

**Comment by Liboul** [comment]
I guess one could try dichotomizing versions:
- No issue on `react-beautiful-dnd`'s latest version
- Issue on `@hello-pangea/dnd`'s latest version

To see which exact version caused the regression, which would narrow down the culprits.
I don't have any time to dedicate to this right now, maybe @Gorbus?

**Comment by kitraining** [comment]
I have the same issue. Is any update regarding the fix?

**Comment by nunocaseiro** [comment]
I have the same issue. Is there any update regarding the fix?

**Comment by Gorbus** [comment]
As react-beautiful-dnd was updated to be react 18 compatible we went back to the other library : [react-beautiful-dnd/](https://github.com/atlassian/react-beautiful-dnd)

**Comment by ghost** [comment]
Is there any updates on this? I am struggling with the same issue...

**Comment by nunocaseiro** [comment]
+1

**Comment by clicktodev** [comment]
> As react-beautiful-dnd was updated to be react 18 compatible we went back to the other library : [react-beautiful-dnd/](https://github.com/atlassian/react-beautiful-dnd)

Are you having the same issue with beautiful dnd?

**Comment by saogregl** [comment]
Is there any updates on this? I am experiencing the same issue... Been thinking about keeping a local state to "isolate" the draggables from the state updates performed by react query, but this requires more (seemingly) unnecessary code.

**Comment by arnolddddddd** [comment]
Also having the same issue react v 18.2.0

**Comment by elsigh** [comment]
I also had the same issue with react-beautiful-dnd. I believe the cause is more wrt react 18 based on all the comments I've read.
Perhaps under the hood this might be a good opportunity to try `useOptimistic`?

**Comment by kr4chinin** [comment]
Still facing this issue, any recent workarounds?

**Comment by danlannan777** [comment]
Also facing this issue. I switched to react-beautiful-dnd and the problem goes away.

**Comment by marcbejar** [comment]
Using an optimistic update with the new reordered array returned from the following function worked for me.

`
const reorder = (list, startIndex, endIndex) => {
  const result = Array.from(list);
  const [removed] = result.splice(startIndex, 1);
  result.splice(endIndex, 0, removed);

  return result;
};
`

Function found on official example: https://codesandbox.io/p/sandbox/k260nyxq9v?file=%2Findex.js%3A13%2C1-19%2C3

**Comment by alantoa** [comment]
I'm also facing this issue, but then I switched to [react-beautiful-dnd](https://github.com/atlassian/react-beautiful-dnd?tab=readme-ov-file) and it works.

**Comment by 55Cancri** [comment]
I think this is a slightly different issue but I also had weird flickering onDragEnd. The issue was that I had a transition css animation defined on the row.

**Comment by mikechabot** [comment]
This is a tricky issue. The only way I've been able to avoid it is to duplicate `react-query` state into `useState`, which is suboptimal, since I now have to duplicate rollback logic within the component as well as `react-query`.

**Comment by janglad** [comment]
Did anyone find a reliable fix for this? Experiencing this right now and it's driving me a bit mad 🫠

**Comment by maxnowack** [comment]
The flickering happens because a re-render occurs after `onDragEnd` is called. In other words, the state update that reflects the new order of sortable items must complete before this re-render. If the state update is asynchronous or deferred to a later tick, the component temporarily renders using the old state, which causes the dropped element to briefly return to its original position.

One solution is to manage the state of the items locally with `useState`, ensuring that the update is applied immediately and then synchronizing with your asynchronous data update.
For example:

```js
const [items, setItems] = useState(initialItems);

const onDragEnd = (result) => {
  if (!result.destination) return;

  const reorderedItems = Array.from(items);
  const [removed] = reorderedItems.splice(result.source.index, 1);
  reorderedItems.splice(result.destination.index, 0, removed);

  // Update local state immediately to prevent flickering.
  setItems(reorderedItems);

  // Optionally, update the remote state asynchronously.
  updateRemoteOrder(reorderedItems).catch((error) => {
    console.error('Remote update failed:', error);
    // Optionally revert to the previous state if needed.
  });
};
```

In my setup, I manage frontend data with [SignalDB](https://signaldb.js.org), which includes an optimistic UI that avoids this flickering issue.

**Comment by cvburgess** [comment]
Watching this issue, will try switching back to the old lib if managing state in two places gets hairy.

**Comment by HendrikGrobler** [comment]
Running into this issue as well, even when using optimistic updates.
Tried messing with the drop animation but didn't have much luck.
I specifically am avoiding keeping local state of the order because it makes other parts harder.
This seems to be the same issue on dnd-kit: https://github.com/clauderic/dnd-kit/issues/921 (they also don't really have a nice solution).

**Comment by lambdaily** [comment]
Any updates on this?

**Comment by statico** [comment]
For now, we worked around this by putting our items in a buffered state which gets updated when the prop changes or when the optimistic update happens.
